File: mergesort.js
function mergesort(a) {
	if (a.length == 1) return a;

	if (a.length == 2) {
		if (a[0] > a[1]) return [a[1], a[0]];
		return a;
	}

	let p = parseInt(a.length / 2, 10);
	let m1 = mergesort(a.slice(0, p));
	let m2 = mergesort(a.slice(p));

	let ret = [];

	do {
		if (m1.length > 0 && m2.length > 0) {
			if (m1[0] <= m2[0]) {
				ret.push(m1[0]);
				m1 = m1.slice(1);
			} else {
				ret.push(m2[0]);
				m2 = m2.slice(1);
			}
		}
		if (m1.length > 0) {
			ret = Array.prototype.push.apply(m1, ret);
			break;
		}
		if (m2.length > 0) {
			ret = Array.prototype.push.apply(m2, ret);
			break;
		}
		break;
	} while (1);

	return ret;
}

a = [5, 9, 1, 3, 4, 6, 6, 3, 2];
console.log(mergesort(a));

File: mergesort.py

def mergesort(a):
    if len(a) == 1:
        return a
    elif len(a) == 2:
        return [a[1], a[0]] if a[0] > a[1] else a

    p = len(a)//2
    m1 = mergesort(a[:p])
    m2 = mergesort(a[p:])

    ret = []

    while True:
        if len(m1) > 0:
            if len(m2) > 0:
                ret.append(m1.pop(0) if m1[0] <= m2[0] else m2.pop(0))
            else:
                ret.extend(m1)
                m1 = []
        elif len(m2) > 0:
            ret.extend(m2)
            m2 = []
        else:
            break
    return ret

a = [5,9,1,3,4,6,6,3,2]
print(mergesort(a))
